src/main/java/com/google/api/codegen/util/ProtoParser.java

```diff
@@ -86,6 +89,51 @@ public ProtoParser(boolean enableProtoAnnotations) {
     }
   }
 
+  /* Return the name of the field representing the header parameter. */
+  public ImmutableSet<String> getHeaderParams(Method method) {
+    ImmutableSet.Builder<String> allParams = ImmutableSet.builder();
+
+    if (!enableProtoAnnotations) {
+      return allParams.build();
+    }
+
+    HttpRule topRule = method.getDescriptor().getMethodAnnotation(AnnotationsProto.http);
+    if (topRule == null) {
+      return allParams.build();
+    }
+    getHeaderParam(topRule).ifPresent(allParams::add);
+
+    // Additional bindings should only be one-deep, according to the API client config spec.
+    // No need to recurse on additional bindings' additional bindings.
+    for (HttpRule rule : topRule.getAdditionalBindingsList()) {
+      getHeaderParam(rule).ifPresent(allParams::add);
+    }
+
+    return allParams.build();
+  }
+
+  // Finds the header param from a HttpRule and add the non-null value to the running set.
+  private Optional<String> getHeaderParam(HttpRule httpRule) {
```

Comment:
Minor, but there is actually nothing wrong in returning `null` especially for the private method. For public method, especially defined on the surface of the library, using `Optional` maybe the right choice, but here `null` is just fine (and technically is slightly more efficient).

Reply:
Ahh, I just didn't feel like doing a null check on a `@Nullable` return value.

I just read https://homes.cs.washington.edu/~mernst/advice/nothing-is-better-than-optional.html, which says that `Optional` is worse than using `@Nullable`, so I'll revert.